prusti-viper/src/encoder/procedure_encoder.rs

```diff
@@ -4463,6 +4624,86 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
     //     // }
     // }
 
+    /// Encode an assignment into an array without an intermediate temporary reference into the
+    /// array.
+    fn encode_array_direct_assign(
+        &mut self,
+        base: PlaceEncoding<'tcx>,
+        index: vir::Expr,
+        array_ty: ty::Ty<'tcx>,
+        rhs: &mir::Rvalue<'tcx>,
+        location: mir::Location,
+    ) -> SpannedEncodingResult<Vec<vir::Stmt>> {
+        let span = self.mir_encoder.get_span_of_location(location);
+
+        // We can't just inhale lookup_pure(base, index) == new_val, because that would
+        // add conflicting info with the previous state of the array at that index.
+        // So the idea is
+        //   label lbl;
+        //   exhale array;
+        //   inhale array;
+        //   // now all info about the array is `havoc`ed
+        //   inhale forall i:: i != index ==> lookup_pure(array, i) == old[lbl](lookup_pure(array, i))
+        //   inhale lookup_pure(array, index) == encoded_rhs
+        //   // now we have all the contents as before, just one item updated
+
+        let (encoded_array, mut stmts) = self.postprocess_place_encoding(
+            base,
+            ArrayAccessKind::Shared,  // shouldn't be nested, so doesn't matter[tm]
+        ).with_span(span)?;
+
+        let label = self.cfg_method.get_fresh_label_name();
+        stmts.push(vir::Stmt::label(&label));
+
+        let at = self.encoder.encode_array_types(array_ty).with_span(span)?;
+
+        let array_acc_expr = vir::Expr::predicate_access_predicate(
+            at.array_pred.clone(),
+            encoded_array.clone(),
+            vir::PermAmount::Write,
+        );
+
+        // exhale and re-inhale to havoc
+        stmts.push(vir!{ exhale [array_acc_expr] });
+        stmts.push(vir!{ inhale [array_acc_expr] });
+
+        let old = |e| { vir::Expr::labelled_old(&label, e) };
+
+        let usize_ty = self.encoder.env().tcx().mk_ty(ty::TyKind::Uint(ty::UintTy::Usize));
+        let idx_val_int = index.field(self.encoder.encode_value_field(usize_ty));
+
+        // inhale infos about array contents back
+        let i_var: vir::Expr = vir_local!{ i: Int }.into();
+        let zero_le_i = vir!{ [vir::Expr::from(0)] <= [ i_var ] };
+        let i_lt_len = vir!{ [ i_var ] < [ vir::Expr::from(at.array_len) ] };
+        let i_ne_idx = vir!{ [ i_var ] != [ old(idx_val_int.clone()) ] };
+        let idx_conditions = vir!{ [zero_le_i] && ([i_lt_len] && [i_ne_idx]) };
```

Comment:
(Not super important, but does the `vir!` macro require this kind of parenthesisation?)

Reply:
yes it does, sadly, as `&&` is a binary thing (i.e. no repeat in and of itself), and the lhs and rhs are `tt` matchers, so we'd match `[zero_le_i] && [i_lt_len]` and then run into "unexpected token" errors.